### DISCUSSION TITLE
Improving performance on Place Order flow and Editor

### DISCUSSION THREAD

**Discussion by wavvves** [woocommerce/woocommerce#46885] [Checkout Flow]
**Summary:** Slow query in `store_has_downloadable_products()` method is affecting performance during template editing and order confirmation.

**Overview**: The method `store_has_downloadable_products()` checks if the store sells downloadable products to determine whether to display the Downloads section during Order Confirmation template editing. However, this method executes a slow query, impacting performance, and runs unnecessarily on every editor call, and with greater severity, during the Place Order flow.
Although it uses a transient to store the result, this may become invalid due to changes in product catalog, triggering unnecessary re-evaluation of the slow query.

Proposed Solutions:

1. Optimize Query Handling: Implement a solution to set the transient once and manage it efficiently upon product creation, deletion, or update.

2. Alternative Approach: Consider bypassing the check altogether and always displaying the Downloads section during template editing, regardless of current downloadable product offerings. This approach avoids transient management overhead and ensures consistent interface behavior.
The Order Confirmation template can be updated to improve clarity around the Downloads section, making it obvious as a section that may not show after placing an order. Also, the user will always be aware of this element when editing the template.

Before:
![image](https://github.com/woocommerce/woocommerce/assets/17236129/32c5ebbe-3a82-4ed5-aaa2-dc7a5aa9690e)

After changing the duplicate “Test Product” to “Test Downloadable Product”. Maybe the word “Test” could be removed altogether 🤔:
![image](https://github.com/woocommerce/woocommerce/assets/17236129/6360bf70-ccd8-46c3-8ca5-120e4f8cec40)

Thoughts? 💡

**Comment by ralucaStan** [comment]
I think the 2nd solution could work with minimum technical effort if it's clear to merchants that the section only shows when there is a downloadable item attached to the order.

But it might add to some confusion on the merchant end, especially if there are customizations added to this page and the store doesn't offer digital assets. Basically, the Editor view of the page would not match the actual confirmation page, which is _mostly_ not aligned with how we handle the other checkout blocks editor: we strive to align the structure of the blocks with what the shopper see.

It would be good to get some input from @pmcpinto and @elizaan36 or @nikkivias on this: would it be ok to show by default the Downloads section, even to stores that don't offer digital goods?

If we decide it's better to show it when it's relevant than I think the first option to use a transient would be the way to go, and would improve the current state of things

**Comment by ralucaStan** [comment]
Looks like we already display the Additional information block in the editor; which is described as `Displays additional information provided by third-party extensions for the current order.`
<img width="1028" alt="Screenshot 2024-04-30 at 17 47 29" src="https://github.com/woocommerce/woocommerce/assets/1628454/40e8ecae-0208-478c-b7ed-b30aacb9e103">

**Comment by elizaan36** [comment]
I'm unable to see the images in the comments, is there any other way to share them?

**Reply by senadir** [reply]
I sent them in slack, not sure why they're not loading for you.

**Reply by elizaan36** [reply]
thank you @senadir !!! I got them.

**Comment by elizaan36** [comment]
> would it be ok to show by default the Downloads section, even to stores that don't offer digital goods?

I don't think we should show the section if the store doesn't offer downloadable goods at all. Is there a check we can do in the store's catalog management settings or something to display the block in the editor view?

For stores that do offer downloadable products, we could take the approach as with custom fields and show a placeholder.
<img width="1007" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5121465/88dcd406-d2d5-437b-a7ca-ff19d20e064f">

Otherwise I think you could use the approach above but remove the "test" as suggested.
<img width="947" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5121465/899b99b3-97ed-44a3-be76-4a2b396e627f">

Noting that the column data should be left aligned except for the far right column. Let me know what you think!

**Comment by pmcpinto** [comment]
> I don't think we should show the section if the store doesn't offer downloadable goods at all. Is there a check we can do in the store's catalog management settings or something to display the block in the editor view?

I agree. Especially considering that stores selling downloadable products are the exception rather than the rule.

**Comment by senadir** [comment]
> I don't think we should show the section if the store doesn't offer downloadable goods at all. Is there a check we can do in the store's catalog management settings or something to display the block in the editor view?

That is the current method of doing things.

@wavvves have you looked into moving the transient calculation logic to action scheduler? if transient is set, use it, otherwise it, otherwise, assume the store does have downloadable and show it.

**Comment by wavvves** [comment]
>@wavvves have you looked into moving the transient calculation logic to action scheduler?

There are some options to better manage the choice to show or hide this, including that. But I don't think it would be ok that that gets updated regularly and blindly, the merchant might create a downloadable product and there's no guarantee the transient will be updated when he tries to update the template.

Since we seem to go the route of keeping the check, maybe the best option would be create it if not existing at first. Then manage it on adding/removing products from catalog. If added virtual set it true, if removed virtual recheck transient (in this case we might use the action scheduler). This would allow to keep the functionality,  using only a one-time expensive query , delegating further need to non critical processes via the action scheduler.

The template section would be readily available for editing after adding virtual products, on the other hand it would depend on the action scheduler before it disappears from template edition, which is less critical.

**Comment by wavvves** [comment]
Issue created https://github.com/woocommerce/woocommerce/issues/47796

Thank you all for your input 🙏
